GeneralisedFilters/src/models/hierarchical.jl

```diff
@@ -29,24 +29,28 @@ function AbstractMCMC.sample(
 )
     outer_dyn, inner_model = model.outer_dyn, model.inner_model
 
-    zs = Vector{eltype(inner_model.dyn)}(undef, T)
-    xs = Vector{eltype(outer_dyn)}(undef, T)
+    zs = OffsetVector(Vector{eltype(inner_model.dyn)}(undef, T + 1), -1)
+    xs = OffsetVector(Vector{eltype(outer_dyn)}(undef, T + 1), -1)
     ys = Vector{eltype(inner_model.obs)}(undef, T)
 
     # Simulate outer dynamics
-    x0 = simulate(rng, outer_dyn; kwargs...)
-    z0 = simulate(rng, inner_model.dyn; new_outer=x0, kwargs...)
+    xs[0] = simulate(rng, outer_dyn; kwargs...)
+    zs[0] = simulate(rng, inner_model.dyn; new_outer=xs[0], kwargs...)
     for t in 1:T
-        prev_x = t == 1 ? x0 : xs[t - 1]
-        prev_z = t == 1 ? z0 : zs[t - 1]
-        xs[t] = simulate(rng, model.outer_dyn, t, prev_x; kwargs...)
+        xs[t] = simulate(rng, model.outer_dyn, t, xs[t - 1]; kwargs...)
```

Comment:
Might be missing something, but why not something like this:
````julia
    xs = Vector{T_dyn}(undef, T)
    ys = Vector{T_obs}(undef, T)

    xs[1] = simulate(rng, model.dyn; kwargs...)
    for t in 1:T
        ys[t] = simulate(rng, model.obs, t, xs[t]; kwargs...)
        if t < T
            xs[t+1] = simulate(rng, model.dyn, t, xs[t]; kwargs...)
        end
    end

Reply:
It's just how we've happened to define the SSMProblems interface. I recall that Hong was keen to have a separation between the initial state x0 and the other states that have associated observations.

Following from that convention, I feel that the offset vector is the cleanest way to achieve it.